src/tasks.ts

```diff
@@ -0,0 +1,331 @@
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
+
+    const feeShares = await Shared.STETH.sharesOf(Shared.CONFIG.CSM_ADDRESS);
+    const distributed = distributeFees(feeShares + 42n);
+
+    const leafs = R.sortBy(R.prop(0), [...Cache.rewards.entries()]);
+    debug('Rewards distribution', leafs);
+    const tree = buildTree(leafs);
+    debug('Report tree root', tree.root);
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+
+    // TODO: Simulate report sending
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
+        targetEpoch,
+        tree.root,
+        distributed,
+        '', // CID TBA, but can be precalculated
+    );
+
+    const cid = await uploadTree(tree);
+    // TODO: send a report
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
+        targetEpoch,
+        tree.root,
+        distributed,
+        cid.toString(),
+    );
+
+    debug('Report sent, storing the artifact');
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+}
+
+async function loadValidators(epoch: Epoch) {
+    const vals = await getModuleValidators(epochLastSlot(epoch));
+    for (const v of vals) {
+        if (v.validator.slashed) {
+            continue;
+        }
+        Cache.indexToPubkey.set(v.index, toHex(v.validator.pubkey));
+    }
+}
+
+async function lookupCommittees(epoch: Epoch) {
+    debug('Looking up committees for epoch', epoch);
+    const epochCommitees = await getEpochCommmittees(epoch);
+    epochCommitees.map((committee) => {
+        committee.validators.forEach((index, pos) => {
+            if (Cache.isTrackedValidator(index)) {
+                Cache.addValidatorToSlotDuty(index, committee.slot, pos);
+            }
+        });
+    });
+}
+
+async function checkEpochSlots(epoch: Epoch) {
+    debug('Checking epoch slots', epoch);
+    for (const slot of iterSlotsInEpoch(epoch)) {
+        const atts = await getSlotAttestations(slot);
+        atts.forEach((a) => {
+            const d = Cache.duties.get(slot);
+            if (d) {
+                d.forEach((c, i) => {
+                    if (a.aggregationBits.get(c.position)) {
+                        const p = Cache.stats.get(
+                            Cache.indexToPubkey.get(c.validatorIndex),
+                        );
+                        if (p) {
+                            p.missedSlots--;
```

Comment:
Can we use another naming? `Missed slot` is already used in context of proposal

Reply:
Can you share a link to the source? I didn't get it.

Reply:
https://www.blocknative.com/blog/missed-slots-and-late-block-reorgs-ethereum
https://www.linkedin.com/pulse/month-were-exploring-slots-anatomy-slot-gossip-layer-issues
https://cryptopotato.com/ethereum-beacon-chains-finality-issues-led-to-253-missed-blocks-in-2-days-glassnode/
https://www.galaxy.com/insights/research/ethereum-consensus-layer-call-93/
https://ethresear.ch/t/cascading-network-effects-on-ethereums-finality/15871
https://etherworld.co/2022/09/22/highlights-of-ethereums-consensus-layer-call-96/
https://blog.bybit.com/en-US/post/market-tumbles-post-merge-ethereum-ushers-in-pos-era-blt7e6b93344f84d2f7/

Reply:
Changed.